xwayland/xwayland.c

```diff
@@ -0,0 +1,227 @@
+	}
+	argv[6] = "-listen";
+	if (asprintf(&argv[7], "%d", wlr_xwayland->x_fd[1]) < 0) {
+		wlr_log_errno(L_ERROR, "asprintf failed");
+		exit(EXIT_FAILURE);
+	}
+	argv[8] = "-wm";
+	if (asprintf(&argv[9], "%d", wlr_xwayland->wm_fd[1]) < 0) {
+		wlr_log_errno(L_ERROR, "asprintf failed");
+		exit(EXIT_FAILURE);
+	}
+
+	const char *xdg_runtime = getenv("XDG_RUNTIME_DIR");
+	if (!xdg_runtime) {
+		wlr_log(L_ERROR, "XDG_RUNTIME_DIR is not set");
+		exit(EXIT_FAILURE);
+	}
+
+	char *envp[3] = { 0 };
+	if (asprintf(&envp[0], "XDG_RUNTIME_DIR=%s", xdg_runtime) < 0 ||
+			asprintf(&envp[1], "WAYLAND_SOCKET=%d", wlr_xwayland->wl_fd[1]) < 0) {
+		wlr_log_errno(L_ERROR, "asprintf failed");
+		exit(EXIT_FAILURE);
+	}
+
+	wlr_log(L_INFO, "Xwayland :%d -rootless -terminate -listen %d -listen %d -wm %d",
+			wlr_xwayland->display, wlr_xwayland->x_fd[0], wlr_xwayland->x_fd[1],
+			wlr_xwayland->wm_fd[1]);
+
+	// TODO: close stdout/err depending on log level
+
+	execvpe("Xwayland", argv, envp);
+}
+
+static bool wlr_xwayland_init(struct wlr_xwayland *wlr_xwayland,
+		struct wl_display *wl_display, struct wlr_compositor *compositor);
+static void wlr_xwayland_finish(struct wlr_xwayland *wlr_xwayland);
+
+static void xwayland_destroy_event(struct wl_listener *listener, void *data) {
+	struct wl_client *client = data;
+	struct wlr_xwayland *wlr_xwayland = wl_container_of(client, wlr_xwayland, client);
+
+	/* don't call client destroy */
+	wlr_xwayland->client = NULL;
+	wlr_xwayland_finish(wlr_xwayland);
+
+	if (wlr_xwayland->server_start - time(NULL) > 5) {
+		wlr_xwayland_init(wlr_xwayland, wlr_xwayland->wl_display,
+				wlr_xwayland->compositor);
+	}
+}
+
+static struct wl_listener xwayland_destroy_listener = {
+	.notify = xwayland_destroy_event,
+};
+
+static void wlr_xwayland_finish(struct wlr_xwayland *wlr_xwayland) {
+
+	if (wlr_xwayland->client) {
+		wl_list_remove(&xwayland_destroy_listener.link);
+		wl_client_destroy(wlr_xwayland->client);
+	}
+
+	if (wlr_xwayland->sigusr1_source) {
+		wl_event_source_remove(wlr_xwayland->sigusr1_source);
+	}
+
+	xwm_destroy(wlr_xwayland->xwm);
+
+	safe_close(wlr_xwayland->x_fd[0]);
+	safe_close(wlr_xwayland->x_fd[1]);
+	safe_close(wlr_xwayland->wl_fd[0]);
+	safe_close(wlr_xwayland->wl_fd[1]);
+	safe_close(wlr_xwayland->wm_fd[0]);
+	safe_close(wlr_xwayland->wm_fd[1]);
+
+	unlink_display_sockets(wlr_xwayland->display);
+	unsetenv("DISPLAY");
+	/* kill Xwayland process? */
+}
+
+static int xserver_handle_ready(int signal_number, void *data) {
+	struct wlr_xwayland *wlr_xwayland = data;
+
+	wlr_log(L_DEBUG, "Xserver is ready");
+
+	wlr_xwayland->xwm = xwm_create(wlr_xwayland);
+	if (!wlr_xwayland->xwm) {
+		wlr_xwayland_finish(wlr_xwayland);
+		return 1;
+	}
+
+	wl_event_source_remove(wlr_xwayland->sigusr1_source);
+	wlr_xwayland->sigusr1_source = NULL;
+
+	char display_name[16];
+	snprintf(display_name, sizeof(display_name), ":%d", wlr_xwayland->display);
+	setenv("DISPLAY", display_name, true);
+
+	return 1;
```

Comment:
this is a wayland event loop processing function, so we return "count" like others which happens to be always 1 here.
To be fair in that case wayland ignores it (it only uses count for 'check' mode), so we could always return 0 if that is clearer